Optimize remove_level_2

romainbrenguier opened the pull request:
Instead of rebuilding the identifier from scratch when we remove the level 2 of the ssa_exprt, we can just strip the suffix, which is more efficient. On benchmarks I ran locally (70 methods from tika) this has improved the execution time of symex by about 8% (from 77.2 sec to 70.8). 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig reviewed (changes requested):
No, this would break any work towards field sensitivity, which `build_ssa_identifier_rec` takes care of.

romainbrenguier commented:
@tautschnig 
> No, this would break any work towards field sensitivity, which build_ssa_identifier_rec takes care of.

What is field sensitivity?

tautschnig commented:
> What is field sensitivity?

Cf. #2574: we would treat individual members of structs or arrays as separate entities in the SSA encoding for the purpose of field-level constant propagation and smaller SAT encodings. Especially for Java it should be highly beneficial. @smowton has poked me about it several times...

romainbrenguier commented:
@tautschnig 
> Cf. #2574: we would treat individual members of structs or arrays as separate entities in the SSA encoding for the purpose of field-level constant propagation and smaller SAT encodings. Especially for Java it should be highly beneficial. @smowton has poked me about it several times...

It sounds interesting but I'm not sure this would be affected by this PR. I will investigate another way to improve how ssa_exprt's are used in symex.

tautschnig commented:
> It sounds interesting but I'm not sure this would be affected by this PR.

The code in `build_ssa_identifier_rec` (which `update_identifier` invokes) would build identifiers like `a!0@1#2.some_member` for a struct member expression `a.some_member`. The change that this PR introduces would turn such an expression into `a!0@1` upon a call to `remove_level_2`, when the correct result is `a!0@1.some_member`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4935a90592cfb82284d7c80d5704f1c016775fd6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94340800

romainbrenguier committed:
Move update_identifier definition to cpp file

To lighten the header.

romainbrenguier committed:
Remove build_identifier from the public_interface

This is not used outside of the ssa_exprt class and shouldn't
be part of the public interface.

romainbrenguier committed:
Document build_ssa_identifier_rec

romainbrenguier committed:
Hide get_l1_name from interface

This is only used in one function.

romainbrenguier committed:
Inline get_l1_name

For ssa_exprt's this always remove_level_2

romainbrenguier commented:
@tautschnig I've changed my approach to optimize the function, this version should be cleaner. I've also added a couple of commits to make the interface of these classes clearer.

smowton commented on src/util/ssa_expr.cpp:
```diff
@@ -82,7 +82,7 @@ bool ssa_exprt::can_build_identifier(const exprt &expr)
     return false;
```
Commit message typo: "and probably shouldn't." -> "and probably shouldn't be part of it."

smowton commented:
Unless I'm misreading this changes very little? Previously inline functions aren't anymore, a couple of functions are static rather than members, and a single "if-is-ssa-expr" test is eliminated? Not objecting, just checking if something has been missed out in shuffling the commits; it currently looks like a cleanup with perhaps 1% performance impact rather than an optimisation.

romainbrenguier commented:
@smowton 
> Unless I'm misreading this changes very little? Previously inline functions aren't anymore, a couple of functions are static rather than members, and a single "if-is-ssa-expr" test is eliminated? Not objecting, just checking if something has been missed out in shuffling the commits; it currently looks like a cleanup with perhaps 1% performance impact rather than an optimisation.

This is mostly true. The essential change is the remove_level_2 simplification. I would expect more than 1% improvement; I will run some benchmarks to confirm that.

romainbrenguier commented:
> I would expect more than 1% improvement; I will run some benchmarks to confirm that.

Yes, a get more than 5% improvement on my benchmarks.

smowton commented:
Just from removing that check? Impressive!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dfd511e5d38a08925711494fcfc254cf6fa8bdbd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94495991

romainbrenguier commented:
@smowton 
> Just from removing that check? Impressive!

It's more than a check because `update_identifier` is rebuilding the identifier strings from scratch, and it's called very frequently during symex.

romainbrenguier commented:
@tautschnig are you happy with the changes?

tautschnig commented on src/util/ssa_expr.cpp:
```diff
@@ -114,5 +114,5 @@ void ssa_exprt::update_identifier()
 {
   remove(ID_L2);
-  update_identifier();
+  set_identifier(get_l1_object_identifier());
```
With this change: shouldn't you move it back to the header file as now everything can be inlined?

tautschnig reviewed (approved):
Ok, looks good to me except there may be further room for improvement by dropping one change as noted below.

romainbrenguier committed:
Optimize remove_level_2

To reset the identifier we can just used the l1_object_identifier which
is a cached version of the level1 identifier.

romainbrenguier commented:
The CI failure is caused by doxygen not being able to be downloaded and installed

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7114c571611e9ab17b8e7bec13e1a8b55248b4e5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94754125
